src/lib.rs

```diff
@@ -252,22 +251,26 @@ impl<'a, T> ExactSizeIterator for FixedSizeListIter<'a, T> {
 }
 
 struct FixedSizeListIterMut<'a, T> {
-    ptr: *mut FixedSizeList<T>,
+    list: &'a mut FixedSizeList<T>,
     front: usize,
     back: usize,
     len: usize,
-    _marker: PhantomData<&'a mut FixedSizeList<T>>,
 }
 
 impl<'a, T> Iterator for FixedSizeListIterMut<'a, T> {
     type Item = &'a mut T;
 
     #[allow(unsafe_code)]
     fn next(&mut self) -> Option<Self::Item> {
-        // Safety: self.ptr has been created from a valid mutable reference
-        let list: &'a mut FixedSizeList<T> = unsafe { &mut *self.ptr };
         if self.len > 0 {
-            let node = list.node_mut(self.front).unwrap();
+            // Safety: This creates a copy of a mutable reference to the list.
+            // This is unsafe in Rust, see "The Rules of References" at
+            // https://doc.rust-lang.org/book/ch04-02-references-and-borrowing.html#the-rules-of-references.
+            // We need to do this because self.list only exists as long as the iterator does,
+            // but the iterator's result items must live longer than the iterator itself.
+            // See https://stackoverflow.com/a/30422716/2013738 for details on reference items outliving iterators.
+            let list_ref = unsafe { &mut *(self.list as *mut FixedSizeList<T>) };
```

Comment:
I wonder if we should instead do this only for the returned `node.data` such that we maintain Rust invariant of never having more than one mutable reference to some data.

Reply:
Yeah, why not. That sounds like a reasonable next step. It might be necessary to do it for `node`, because we access that locally.

The idea here is to better understand what happened before. From there I'm sure further implrovements can be done.

Reply:
Sure! Could you rebase then?